### PR TITLE
Implement SP App deletion on Tenant Deletion

### PR DESCRIPTION
**Pull request by sumedhe**
### Proposed changes in this pull request

Implemented the deletion of Service Provider Apps when a Tenant gets deleted

Fixes: #2932

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java [hunk 1]
```diff
@@ -131,6 +131,14 @@ public abstract void updateApplication(ServiceProvider serviceProvider, String t
     public abstract void deleteApplication(String applicationName, String tenantDomain, String username)
             throws IdentityApplicationManagementException;
 
+    /**
+     * Delete Applications by tenant id.
+     *
+     * @param tenantId Id of the tenant
+     * @throws IdentityApplicationManagementException
+     */
+    public abstract void deleteApplicationsByTenantId(int tenantId) throws IdentityApplicationManagementException;
```
Comment (IsuraD) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java:140]: rename to deleteApplications(int tenantId)

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 2]
```diff
@@ -714,6 +714,20 @@ public void deleteApplication(String applicationName, String tenantDomain, Strin
...
+     * @throws IdentityApplicationManagementException
+     */
+    @Override
+    public void deleteApplicationsByTenantId(int tenantId) throws IdentityApplicationManagementException {
+
+        ApplicationDAO appDAO = ApplicationMgtSystemConfig.getInstance().getApplicationDAO();
+        appDAO.deleteApplicationsByTenantId(tenantId);
+    }
+
+
```
Comment (IsuraD) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:730]: remove line

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 3]
```diff
@@ -3538,6 +3540,41 @@ public void deleteApplication(int applicationID, Connection connection)
...
+    /**
+     * Delete all applications of a given tenant id.
+     *
+     * @param
+     * @throws IdentityApplicationManagementException
+     */
+    @Override
+    public void deleteApplicationsByTenantId(int tenantId) throws IdentityApplicationManagementException {
+
+        if (log.isDebugEnabled()) {
```
Comment (IsuraD) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3552]: Lets put an audit log here...

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java (hunk 1 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java:140]: 👍

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java (hunk 2 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:730]: 👍

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 3 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3552]: 👍

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 4]
```diff
@@ -3538,6 +3545,45 @@ public void deleteApplication(int applicationID, Connection connection)
 
     }
 
+    /**
+     * Delete all applications of a given tenant id.
+     *
+     * @param
```
Comment (ShanChathusanda93) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3551]: Add the params of the method.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 5]
```diff
@@ -3538,6 +3545,45 @@ public void deleteApplication(int applicationID, Connection connection)
...
     }
 
+    /**
+     * Delete all applications of a given tenant id.
+     *
+     * @param tenantId Id of the tenant
+     * @throws IdentityApplicationManagementException
+     */
+    @Override
+    public void deleteApplicationsByTenantId(int tenantId) throws IdentityApplicationManagementException {
```
Comment (IsuraD) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3555]: rename to  public void deleteApplications(int tenantId)

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationTenantMgtListener.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import org.wso2.carbon.identity.application.common.IdentityApplicationManagementException;
+import org.wso2.carbon.identity.application.mgt.ApplicationManagementService;
+import org.wso2.carbon.stratos.common.beans.TenantInfoBean;
+import org.wso2.carbon.stratos.common.exception.StratosException;
+import org.wso2.carbon.stratos.common.listeners.TenantMgtListener;
+
+/**
+ * Application Tenant Management Listener
+ */
+public class ApplicationTenantMgtListener implements TenantMgtListener {
```
Comment (IsuraD) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationTenantMgtListener.java:12]: Use org.wso2.carbon.identity.oauth2.listener.TenantCreationEventListener

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 7]
```diff
@@ -3538,6 +3545,45 @@ public void deleteApplication(int applicationID, Connection connection)
...
+        if (log.isDebugEnabled()) {
+            log.debug("Deleting all applications of the tenant: " + tenantId);
+        }
+
+        String auditData = "\"" + "Tenant Id" + "\" : \"" + tenantId + "\"";
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+        PreparedStatement deleteClientPrepStmt = null;
+
+        try {
```
Comment (IsuraD) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3566]: better use try with resource

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationTenantMgtListener.java (hunk 6 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationTenantMgtListener.java]: Removed `ApplicationTenantMgtListener` and will use `org.wso2.carbon.identity.oauth2.listener.TenantCreationEventListener` instead.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java [hunk 8]
```diff
@@ -131,6 +131,14 @@ public abstract void updateApplication(ServiceProvider serviceProvider, String t
     public abstract void deleteApplication(String applicationName, String tenantDomain, String username)
             throws IdentityApplicationManagementException;
 
+    /**
+     * Delete Applications by tenant id.
+     *
+     * @param tenantId Id of the tenant
+     * @throws IdentityApplicationManagementException
```
Comment (sachiniWettasinghe) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java:138]: Complete the Javadoc comment by adding the specific Exception type. The param descriptions should start with a capital letter and end with a full stop.

Comment (sachiniWettasinghe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java:138]: Check every other instance and fix this

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 9]
```diff
@@ -714,6 +718,38 @@ public void deleteApplication(String applicationName, String tenantDomain, Strin
...
+    public void deleteApplications(int tenantId) throws IdentityApplicationManagementException {
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomain(tenantId);
+        ApplicationBasicInfo[] applicationBasicInfos = getAllApplicationBasicInfo(
+                tenantDomain, CarbonContext.getThreadLocalCarbonContext().getUsername());
+
+        ApplicationDAO appDAO = ApplicationMgtSystemConfig.getInstance().getApplicationDAO();
+        appDAO.deleteApplications(tenantId);
+
+        // Clear cache entries of each deleted SP
```
Comment (sachiniWettasinghe) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:737]: Comment should end with a full stop.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 10]
```diff
@@ -3538,6 +3545,41 @@ public void deleteApplication(int applicationID, Connection connection)
...
+
+        if (log.isDebugEnabled()) {
+            log.debug("Deleting all applications of the tenant: " + tenantId);
+        }
+
+        String auditData = "\"" + "Tenant Id" + "\" : \"" + tenantId + "\"";
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(true)) {
+
+            // Delete the application certificates of the tenant
```
Comment (sachiniWettasinghe) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3565]: end the comment with a full stop

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 11]
```diff
@@ -5060,4 +5125,24 @@ private int createServiceProvider(String tenantDomain, ServiceProvider servicePr
         }
         return applicationId;
     }
+
+    /**
+     * Add audit log entry.
+     *
+     * @param action Action
+     * @param data Audit data
+     * @param result Result
```
Comment (sachiniWettasinghe) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:5134]: Complete the Javadoc descriptions

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 12]
```diff
@@ -5060,4 +5125,24 @@ private int createServiceProvider(String tenantDomain, ServiceProvider servicePr
...
+     */
+    private void audit(String action, String data, String result) {
+
+        String loggedInUser = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername();
+        if (StringUtils.isBlank(loggedInUser)) {
+            loggedInUser = CarbonConstants.REGISTRY_SYSTEM_USERNAME;
+        }
+
```
Comment (sachiniWettasinghe) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:5142]: Remove redundant line

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java (hunk 8 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementService.java:138]: Fixed in #3312

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java (hunk 9 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:737]: Fixed in #3312

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 10 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:3565]: Fixed in #3312

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 11 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:5134]: Fixed in #3312

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 12 above)
Comment (sumedhe) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:5142]: Fixed in #3312

**Pull request merged**